sqlmesh/core/engine_adapter/base.py

```diff
@@ -748,6 +749,153 @@ def _merge(
             )
         )
 
+    def scd_type_2(
+        self,
+        target_table: TableName,
+        source_table: QueryOrDF,
+        unique_key: t.Sequence[str],
+        valid_from_name: str,
+        valid_to_name: str,
+        updated_at_name: str,
+        execution_time: TimeLike,
+        columns_to_types: t.Optional[t.Dict[str, exp.DataType]] = None,
+        **kwargs: t.Any,
+    ) -> None:
+        if columns_to_types is None:
+            columns_to_types = self.columns(target_table)
+        if updated_at_name not in columns_to_types:
+            raise SQLMeshError(
+                f"Column {updated_at_name} not found in {target_table}. Table must contain an `updated_at` timestamp for SCD Type 2"
+            )
+        unmanaged_columns = [
+            col for col in columns_to_types if col not in {valid_from_name, valid_to_name}
+        ]
+        df = self.try_get_pandas_df(source_table)
+        if df is not None:
+            source_table = next(
+                pandas_to_sql(
+                    df,
+                    columns_to_types={
+                        k: v for k, v in columns_to_types.items() if k in unmanaged_columns
+                    },
+                )
+            )
+        start_time = "CAST('1970-01-01 00:00:00' AS TIMESTAMP)"
+        query = (
+            exp.Select()  # type: ignore
+            .with_(
+                "source",
+                exp.select(*unmanaged_columns)
+                .distinct(*unique_key)
```

Comment:
is this distinct(...) supported in most dialects?

Reply:
I actually stole this from logical merge mixin: https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/core/engine_adapter/mixins.py#L53

I assumed since it was there that it was considered safe. I can rewrite both if we aren't confident.

Reply:
We tried `distinct` on 4 or 5 of the most used dialects when logical merge was in development. It was supported in all of them.

Reply:
sounds good